modules/core/src/main/java/org/apache/ignite/internal/binary/builder/BinaryBuilderSerializer.java

```diff
@@ -141,6 +141,11 @@ public void writeValue(BinaryWriterExImpl writer, Object val, boolean forceCol,
         if (forceCol || BinaryUtils.isSpecialCollection(val.getClass())) {
             Collection<?> c = (Collection<?>)val;
 
+            if (writer.tryWriteAsHandle(c))
+                return;
+
+            writer.out().unsafeEnsure(1 + 4 + 1);
```

Comment:
Why check capacity should be used only here? Please remove or add for each writer.writeXxx() places.